### ISSUE TITLE
RokBooster improvements for G5

### ISSUE THREAD

**Issue by JoomFX** [gantry/gantry5#1006]
@w00fz @mahagr Hey guys,

I know that this request is supposed to be raised in the RT Forums, but I just have the feeling that here you will pay more attention to it :smile: 

So, here are the improvements I'd like to ask you to implement in RokBooster... no, no, I'm begging you to implement those :smile: 

1. Combine (and compress) the scripts loaded in the `footer` as well (`{% scripts in 'footer' with { priority: 10 } %}`). Currently, it seems that RokBooster only looks for scripts in the `<head>` but in G5 we have `main.js` in the `footer` (I also have couple other custom scripts in the `footer`). 
If you really want to keep some scripts (`main.js`) in the footer, can you then combine (and compress) all scripts from the `footer` into a second PHP file?

2. "Defer parsing of JavaScript" - I'm not really sure about this one. Implementing this (adding `async`) might cause more bad than good, but still it might be worth trying. Just as an example, JCH Optimize has this option, to defer the JS. 

Basically, only these two things are currently missing in RokBooster.
Thank you in advance ;)

**Comment by JoomFX** [comment]
Something interesting... I just did a test with a template built on Helix 3 - I installed and configured RokBooster there and when I checked it in GTMetrix it reported "Defer parsing of JavaScript" as 100% :open_mouth: 

It means that in Helix 3 they are handling this somehow.

**Comment by ramon12** [comment]
Defer means postpone, load after initial load of page. Eg. Load html first, then css, and then js. You can defer all external resources. You can do that with a script to set the loading order.

**Comment by JoomFX** [comment]
Thanks for the explanation @ramon12 ... yep, I know what defer means :)

> You can defer all external resources. You can do that with a script to set the loading order.

Yes, and the thing I'm asking in this issue is, to add such functionality in RokBooster, instead of using "a script to set the loading order".

JCH Optimize has a "Defer parsing of JavaScript" functionality. There are also [other extensions](http://extensions.joomla.org/extension/javascript-async-and-defer) that offer this.

In this issue I'm also asking about another thing, specifically related to G5. To be more precise - Combine (and compress) the scripts loaded in the `footer`.

**Comment by w00fz** [comment]
The defer is something I tried real quick to add to rokbooster not so long ago and have had troubles. It requires a bit more digging as we need to make sure we group right the deferred js from the non deferred, rather than all together.
We definitely need to look into it, as well as compressing the footer js. I'd also like to have an option to decide wheter the compressed js should render in footer rather than head.

Thanks for bringing this up.

// Djamil Legato // From iPhone //

**Comment by JoomFX** [comment]
@w00fz Thanks Djamil! I'm glad you are considering this and I'm looking forward for the implementation :)

With these two improvements, a G5-based website (on a properly configured/optimized server) might have GTMetrix results close to 100% :smile:

**Comment by mark118** [comment]
Why does Rokbooster create php files instead of css/js files? Does that impact gzip/caching?

**Comment by w00fz** [comment]
When you reach the `.php` file, the file in question is served as proper CSS or JS content, even if the extension is php.

The reason it generates PHP files is to inject PHP code that generates optimized expiration headers for the file (things like `Expires` and `Cache-Control`), making it de-facto a dynamic asset (CSS/JS).

It does impact gzip/caching but not in a bad way, it actually impacts it in a way that makes it more optimized for gzip and caching.

**Comment by mark118** [comment]
Ok that's clear, thank you. I'll have to see what Cloudflare does with it :-)
